zcash_client_sqlite/src/wallet/init/migrations/fix_bad_change_flagging.rs

```diff
@@ -0,0 +1,198 @@
+//! Sets the `is_change` flag on output notes received by an internal key when input value was
+//! provided from the account corresponding to that key.
+use std::collections::HashSet;
+
+use rusqlite::named_params;
+use schemerz_rusqlite::RusqliteMigration;
+use uuid::Uuid;
+use zip32::Scope;
+
+use crate::{
+    wallet::{
+        init::{migrations::fix_broken_commitment_trees, WalletMigrationError},
+        scope_code,
+    },
+    SAPLING_TABLES_PREFIX,
+};
+
+#[cfg(feature = "orchard")]
+use crate::ORCHARD_TABLES_PREFIX;
+
+pub(super) const MIGRATION_ID: Uuid = Uuid::from_u128(0x6d36656d_533b_4b65_ae91_dcb95c4ad289);
+
+const DEPENDENCIES: &[Uuid] = &[fix_broken_commitment_trees::MIGRATION_ID];
+
+pub(super) struct Migration;
+
+impl schemerz::Migration<Uuid> for Migration {
+    fn id(&self) -> Uuid {
+        MIGRATION_ID
+    }
+
+    fn dependencies(&self) -> HashSet<Uuid> {
+        DEPENDENCIES.iter().copied().collect()
+    }
+
+    fn description(&self) -> &'static str {
+        "Sets the `is_change` flag on output notes received by an internal key when input value was provided from the account corresponding to that key."
+    }
+}
+
+impl RusqliteMigration for Migration {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &rusqlite::Transaction) -> Result<(), WalletMigrationError> {
+        let fix_change_flag = |table_prefix| {
+            transaction.execute(
+                &format!(
+                    "UPDATE {table_prefix}_received_notes
+                     SET is_change = 1
+                     FROM sent_notes sn
+                     WHERE sn.tx = {table_prefix}_received_notes.tx
+                     AND sn.from_account_id = {table_prefix}_received_notes.account_id
+                     AND {table_prefix}_received_notes.recipient_key_scope = :internal_scope"
```

Comment:
This restriction means that funds sent from an account to itself via its external address will not be updated here. For modern wallets that makes sense, but for historic wallets (when internal scope didn't exist) a "change" output was simply an output received by the account in a tx that the account also spent funds in. Do we have reason to believe that those earlier transactions are or are not affected by this bug?

Reply:
The legacy shielding transactions likely _will_ be affected by this problem in wallet recovery. However, I think that's acceptable; the wallets at the time those transactions were created did not show "shielding" for those transactions in wallet history, so this leaves that part of the transaction history essentially unmodified from the user's perspective.